### PR TITLE
retry on rate limit errors 

### PR DESCRIPTION
**Pull request by smola**
Scaleway API has a rate limit of 100 requests per second. When the limit is reached, the API replies with 429 HTTP status.

So, for every API request, we retry with exponential backoff whenever we find a 429 error.

### CODE REVIEW

**Comment by smola** [comment]
This PR is stacked on top of https://github.com/terraform-providers/terraform-provider-scaleway/pull/29
I'll rebase once https://github.com/terraform-providers/terraform-provider-scaleway/pull/29 is merged.

**Comment by nicolai86** [comment]
Thank you @smola . I’ll take a look today!

File: scaleway/resource_volume_attachment.go [hunk 1]
```diff
@@ -214,31 +229,19 @@ func resourceScalewayVolumeAttachmentDelete(d *schema.ResourceData, m interface{
 		volumes[k] = v
 	}
 
-	if err := resource.Retry(serverWaitTimeout, func() *resource.RetryError {
+	if err := retryWithCodes([]int{429, 400}, func() error {
```
Comment (nicolai86) [inline scaleway/resource_volume_attachment.go]: I don't think we should treat 400 - BadRequest - as retryable here.

File: scaleway/resource_volume_attachment.go [hunk 2]
```diff
@@ -90,31 +100,19 @@ func resourceScalewayVolumeAttachmentCreate(d *schema.ResourceData, m interface{
...
```
Comment (nicolai86) [inline scaleway/resource_volume_attachment.go]: Why treat 400 as retryable?

File: scaleway/helpers.go [hunk 3]
```diff
@@ -117,3 +126,27 @@ func waitForServerState(scaleway *api.API, serverID, targetState string) error {
...
 	return err
 }
+
+func retry(f func() error) error {
+	return retryWithCodes([]int{429}, f)
+}
+
+func retryWithCodes(retryCodes []int,
+	f func() error) error {
+	return resource.Retry(serverWaitTimeout, func() *resource.RetryError {
```
Comment (nicolai86) [inline scaleway/helpers.go]: Please introduce a new constant for this - `serverWaitTimeout` is specific to the boot time of scaleway servers.

**Comment by nicolai86** [comment]
the tests are green: 

```
?   	github.com/terraform-providers/terraform-provider-scaleway	[no test files]
=== RUN   TestAccScalewayDataSourceBootscript_Filtered
--- PASS: TestAccScalewayDataSourceBootscript_Filtered (4.42s)
=== RUN   TestAccScalewayDataSourceImage_Basic
--- PASS: TestAccScalewayDataSourceImage_Basic (11.09s)
=== RUN   TestAccScalewayDataSourceImage_Filtered
--- PASS: TestAccScalewayDataSourceImage_Filtered (9.92s)
=== RUN   TestAccScalewayIP_importBasic
--- PASS: TestAccScalewayIP_importBasic (3.30s)
=== RUN   TestAccScalewaySecurityGroup_importBasic
--- PASS: TestAccScalewaySecurityGroup_importBasic (3.87s)
=== RUN   TestAccScalewayServer_importBasic
--- PASS: TestAccScalewayServer_importBasic (124.31s)
=== RUN   TestAccScalewayVolume_importBasic
--- PASS: TestAccScalewayVolume_importBasic (2.32s)
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestProvider_impl
--- PASS: TestProvider_impl (0.00s)
=== RUN   TestAccScalewayIP_Count
--- PASS: TestAccScalewayIP_Count (5.59s)
=== RUN   TestAccScalewayIP_Basic
--- PASS: TestAccScalewayIP_Basic (13.21s)
=== RUN   TestAccScalewaySecurityGroupRule_Basic
--- PASS: TestAccScalewaySecurityGroupRule_Basic (8.42s)
=== RUN   TestAccScalewaySecurityGroupRule_Count
--- PASS: TestAccScalewaySecurityGroupRule_Count (9.05s)
=== RUN   TestAccScalewaySecurityGroup_Basic
--- PASS: TestAccScalewaySecurityGroup_Basic (4.87s)
=== RUN   TestAccScalewayServer_Basic
--- PASS: TestAccScalewayServer_Basic (218.31s)
=== RUN   TestAccScalewayServer_ExistingIP
--- PASS: TestAccScalewayServer_ExistingIP (139.46s)
=== RUN   TestAccScalewayServer_Volumes
--- PASS: TestAccScalewayServer_Volumes (138.71s)
=== RUN   TestAccScalewayServer_SecurityGroup
--- PASS: TestAccScalewayServer_SecurityGroup (159.95s)
=== RUN   TestAccScalewayVolumeAttachment_Basic
--- PASS: TestAccScalewayVolumeAttachment_Basic (433.79s)
=== RUN   TestAccScalewayVolume_Basic
--- PASS: TestAccScalewayVolume_Basic (2.90s)
PASS
```

once we clarify the above comments we're good to merge!

**Pull request closed without merging**